controllers/replica_calculator.go

```diff
@@ -212,62 +215,117 @@ func (c *ReplicaCalculator) GetResourceReplicas(logger logr.Logger, target *auto
+	setCondition(wpa, v1alpha1.WatermarkPodAutoscalerStatusConvergeToHighWatermark, corev1.ConditionFalse, convergeToHighWatermarkAllowedMessage, convergeToHighWatermarkAllowedReason)
 
-		setCondition(wpa, v1alpha1.WatermarkPodAutoscalerStatusAboveHighWatermark, corev1.ConditionFalse, aboveHighWatermarkReason, aboveHighWatermarkAllowedMessage)
-		setCondition(wpa, v1alpha1.WatermarkPodAutoscalerStatusBelowLowWatermark, corev1.ConditionTrue, belowLowWatermarkReason, belowLowWatermarkAllowedMessage)
+	return replicaCount
+}
 
-	default:
-		restrictedScaling.With(labelsWithReason).Set(1)
-		value.With(labelsWithMetricName).Set(adjustedUsage)
-		logger.Info("Within bounds of the watermarks", "value", utilizationQuantity.String(), "currentReadyReplicas", currentReadyReplicas, "tolerance (%)", float64(wpa.Spec.Tolerance.MilliValue())/10, "adjustedLM", adjustedLM, "adjustedHM", adjustedHM, "adjustedUsage", adjustedUsage)
+func getReplicaCountWithinBounds(currentReplicas int32, wpa *v1alpha1.WatermarkPodAutoscaler) (replicaCount int32) {
+
+	setCondition(wpa, v1alpha1.WatermarkPodAutoscalerStatusAboveHighWatermark, corev1.ConditionFalse, aboveHighWatermarkReason, aboveHighWatermarkAllowedMessage)
+	setCondition(wpa, v1alpha1.WatermarkPodAutoscalerStatusBelowLowWatermark, corev1.ConditionFalse, belowLowWatermarkReason, belowLowWatermarkAllowedMessage)
+	setCondition(wpa, v1alpha1.WatermarkPodAutoscalerStatusConvergeToHighWatermark, corev1.ConditionFalse, convergeToHighWatermarkAllowedMessage, convergeToHighWatermarkAllowedReason)
 
-		setCondition(wpa, v1alpha1.WatermarkPodAutoscalerStatusAboveHighWatermark, corev1.ConditionFalse, aboveHighWatermarkReason, aboveHighWatermarkAllowedMessage)
-		setCondition(wpa, v1alpha1.WatermarkPodAutoscalerStatusBelowLowWatermark, corev1.ConditionFalse, belowLowWatermarkReason, belowLowWatermarkAllowedMessage)
+	// returning the currentReplicas instead of the count of healthy ones to be consistent with the upstream behavior.
+	return currentReplicas
+}
+
+// tryToConvergeToHw will try to make the replicaCount slowly converge to high watermark but make sure that it does
+// not trigger a downscale that would put usage above HW
+func tryToConvergeToHw(logger logr.Logger, currentReplicas, currentReadyReplicas int32, wpa *v1alpha1.WatermarkPodAutoscaler, adjustedUsage float64, lowMark, highMark *resource.Quantity) (replicaCount int32) {
+
+	downScaleBy := *wpa.Spec.ReplicaScalingAbsoluteModulo
+	currentReplicasAfterDownscale := math32.Max(currentReplicas-downScaleBy, 1)
+	currentReadyReplicasAfterDownscale := math32.Max(currentReadyReplicas-downScaleBy, 0)
 
-		// returning the currentReplicas instead of the count of healthy ones to be consistent with the upstream behavior.
-		return currentReplicas, utilizationQuantity.MilliValue()
+	adjustedUsageAfterDownscale := math.Ceil(adjustedUsage * float64(currentReadyReplicas) / float64(currentReadyReplicasAfterDownscale))
```

Comment:
if downscale by is > currentReadyReplicas, currentReadyReplicasAfterDownscale would be 0 or am I missing something?

Reply:
That's totally correct, we never want `currentReadyReplicasAfterDownscale` to be 0 anyway, the Max should be `1` too